pkg/aws/services/vpclattice.go

```diff
@@ -54,125 +80,101 @@ func NewDefaultLattice(sess *session.Session, region string) *defaultLattice {
 
 func (d *defaultLattice) ListServiceNetworksAsList(ctx context.Context, input *vpclattice.ListServiceNetworksInput) ([]*vpclattice.ServiceNetworkSummary, error) {
 	result := []*vpclattice.ServiceNetworkSummary{}
-	resp, err := d.ListServiceNetworksWithContext(ctx, input)
 
-	for {
-		if err != nil {
-			return nil, err
-		}
-		for _, mesh := range resp.Items {
-			result = append(result, mesh)
-		}
-		if resp.NextToken == nil {
-			break
+	err := d.ListServiceNetworksPagesWithContext(ctx, input, func(page *vpclattice.ListServiceNetworksOutput, lastPage bool) bool {
+		for _, sn := range page.Items {
+			result = append(result, sn)
 		}
-		input.NextToken = resp.NextToken
-		resp, err = d.ListServiceNetworksWithContext(ctx, input)
+		return true
+	})
```

Comment:
I've updated all the methods here to use automatic pagination as per @mikhail-aws suggestion in previous PR. The way these methods work is you provide a function for them to call for each page found and you return ```true``` if you want the call to keep fetching pages.

Reply:
compact version:

```go
 7 func (d *defaultLattice) ListServiceNetworksAsList(ctx context.Context, input *vpclattice.ListServiceNetworksInput) ([]*vpclattice.ServiceNetworkSummary, error) {
  8         result := []*vpclattice.ServiceNetworkSummary{}
  9         err := d.ListServiceNetworksPagesWithContext(ctx, input, func(page *vpclattice.ListServiceNetworksOutput, lastPage bool) bool {
 10                 result = append(result, page.Items...)
 11         }
 12         return result, err
 13 }
```

Generic get all pages function. Since every page is a struct, getting Items from page can be done through reflection, or explicitly written as getItemsFn, that return items from page.

```go
 func ListAll[REQ any, ITEM any, PAGE any](
         ctx context.Context,    req REQ,
         listPages func(context.Context, REQ, func(PAGE, bool) bool, ...request.Option) error,
         getPageItems func(PAGE) []ITEM)([]ITEM, error) {
                 out := []ITEM{}
                 err := listPages(ctx, req, func(page PAGE, _ bool) bool {
                         out = append(out, getPageItems(page)...)
                         return true
                 })
                 return out, err
         }

 func (l *Lattice) ListServiceNetworksAsList(ctx context.Context, in *vpclattice.ListServiceNetworksInput) ([]*vpclattice.ServiceNetworkSummary, error) {
         return ListAll(
                 ctx, in, l.c.ListServiceNetworksPagesWithContext,
                 func(page *vpclattice.ListServiceNetworksOutput) []*vpclattice.ServiceNetworkSummary {
                         return page.Items
                 })
 }
```

Reply:
I dont see any value in unit-tests with mocks for these Lattice wrapper functions. As you mentioned, it's painful and kind of self-fulfilling prophecy.

Reply:
I am curious what the benefit of  `ListServiceNetworksPagesWithContext(ctx, input, fn)`  compared to the old `NextToken` way? (Still learning this part)

Reply:
Honestly I'm not convinced it's that big a win. The logic is a little harder to wrap your head around, but it means we can't get into an infinite loop or accidently omit pages by mishandling ```NextToken```.